### PR TITLE
Initial conference paper bibliography consistency check

### PR DESCRIPTION
**Pull request by koppor**
When writing a paper, one copies BibTeX entries to the .bib file for the paper. Then, one deletes fields to make the library shorter. Maybe, one does not touch the entries at all. All in all, there will be inconsistencies in the bib file, which typically need manual investigation. (More discussion on that topic at https://github.com/JabRef/jabref/issues/160)

This PR adds the logic to check the references of a given list of BibEntries for consistency: Do the entries of the same entry type have the same fields set?

The output is a CSV file.

For a real-life publication (pseudonymized here based on the [from-aux functionality](https://docs.jabref.org/advanced/newbasedonaux)), the output is as follows

image

---

Future work is to provide an integration in the UI.

---

This PR gets reviewable if https://github.com/JabRef/jabref/pull/10776 is merged. Therefore marked as draft.

### Mandatory checks

- [ ] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ ] Tests created for changes (if applicable)
- [ ] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [ ] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

File: src/main/java/org/jabref/logic/quality/consistency/PaperConsistencyCheckResultCsvWriter.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+                              .map(BibEntryType::getOptionalFields)
+                              .stream()
+                              .flatMap(bibFieldSet -> bibFieldSet.stream())
+                              .map(BibField::field)
+                              .collect(Collectors.toSet());
+
+                      PaperConsistencyCheck.EntryTypeResult entries = mapEntry.getValue();
+                      SequencedCollection<BibEntry> bibEntries = entries.sortedEntries();
+
+                      bibEntries.forEach(Unchecked.consumer(bibEntry -> {
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/logic/quality/consistency/PaperConsistencyCheckResultCsvWriter.java:91]: Refactor to own method

File: src/main/java/org/jabref/logic/quality/consistency/PaperConsistencyCheckResultCsvWriter.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+                          List<String> theRecord = new ArrayList(columnCount);
+                          theRecord.add(entryType);
+                          theRecord.add(bibEntry.getCitationKey().orElse(""));
+                          allFields.forEach(field -> {
+                              theRecord.add(bibEntry.getField(field).map(value -> {
+                                  if (requiredFields.contains(field)) {
+                                      return "x";
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/logic/quality/consistency/PaperConsistencyCheckResultCsvWriter.java:98]: make this constants

File: src/main/java/org/jabref/logic/quality/consistency/PaperConsistencyCheck.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+            if (uniqueFields.isEmpty()) {
+                return;
+            }
+
+            List<BibEntry> sortedEntries = entryTypeToEntriesMap
+                    .get(entryType).stream()
+                    .filter(entry -> !entry.getFields().equals(commonFields))
+                    .sorted((e1, e2) -> {
+                        int first = e1.getFields().size() - e2.getFields().size();
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/logic/quality/consistency/PaperConsistencyCheck.java:73]: Extract this to a comparator method stuff

**Review by Siedlerchr — CHANGES_REQUESTED** [review]
some refactorings necessary

**Comment by koppor** [comment]
Will try to add MArkdown output in this PR, too.

**Comment by Siedlerchr** [comment]
Should we merge this first draft?

**Comment by koppor** [comment]
> Should we merge this first draft?

Very OK for me!

There won't be markdown, but an ASCII ouput (like the one of [lychee](https://github.com/lycheeverse/lychee/)) to really enable it for usage in a CI.

**Review by Siedlerchr — APPROVED** [review]

**Pull request merged**